sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -533,6 +538,56 @@ public static Unbound withSideInputs(
     return DisplayData.item("fn", fn.getClass()).withLabel("Transform Function");
   }
 
+  private static void finishSpecifyingStateSpecs(
+      DoFn<?, ?> fn,
+      CoderRegistry coderRegistry,
+      Coder<?> inputCoder) {
+    DoFnSignature signature = DoFnSignatures.getSignature(fn.getClass());
+    Map<String, DoFnSignature.StateDeclaration> stateDeclarations = signature.stateDeclarations();
+    for (DoFnSignature.StateDeclaration stateDeclaration : stateDeclarations.values()) {
+      try {
+        StateSpec<?, ?> stateSpec = (StateSpec<?, ?>) stateDeclaration.field().get(fn);
+        stateSpec.offerCoders(codersForStateSpecTypes(stateDeclaration, coderRegistry, inputCoder));
+        stateSpec.finishSpecifying();
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      } catch (Exception e) {
+        throw new Pipeline.PipelineExecutionException(e);
+      }
+    }
+  }
+
+  /**
+   * Try to provide coders for as many of the type arguments of given
+   * {@link DoFnSignature.StateDeclaration} as possible.
+   */
+  private static <InputT> Coder[] codersForStateSpecTypes(
+      DoFnSignature.StateDeclaration stateDeclaration,
+      CoderRegistry coderRegistry,
+      Coder<InputT> inputCoder) {
+    Type stateType = stateDeclaration.stateType().getType();
+    if (stateType instanceof ParameterizedType) {
```

Comment:
readability nit: I'd reverse the branches here to put the short one first, and just have a short circuit `if not ParamterizedType return new Coder[0]`. Also saves some indentation.